Source/WebDriver/WebSocketServer.cpp

```diff
@@ -0,0 +1,234 @@
+    // let listener be that listener. Otherwise let listener be a new WebSocket listener with implementation-defined
+    // host, port, secure flag, and an empty list of WebSocket resources.
+    // Implementation note: As we support only one session at a time, we always reuse the same listener
+    auto listener = m_listener;
+
+    // 2. Let resource name be the result of constructing a WebSocket resource name given session.
+    auto resourceName = constructWebSocketResourceNameGivenSession(sessionId);
+
+    // 3. Append resource name to the list of WebSocket resources for listener.
+    listener->resources.push_back(resourceName);
+
+    // 4. Append listener to the remote end's active listeners.
+    // Implementation note: We don't have a list of active listeners, as we support only one session at a time.
+
+    // 5. Return listener.
+    return listener;
+}
+
+String WebSocketServer::constructWebSocketResourceNameGivenSession(String sessionId)
+{
+    // https://w3c.github.io/webdriver-bidi/#construct-a-websocket-resource-name
+
+    // 1. If session is null, return "/session"
+    if (sessionId.isNull())
+        return "/session"_s;
+
+    // 2. Return the result of concatenating the string "/session/" with session’s session ID.
+    return makeString("/session/"_s, sessionId);
+}
+
+String WebSocketServer::constructWebSocketURLForListenerAndSession(const RefPtr<WebSocketListener> listener, String sessionId)
+{
+    // https://w3c.github.io/webdriver-bidi/#construct-a-websocket-url
+    // Let resource name be the result of construct a WebSocket resource name with session.
+    auto resourceName = constructWebSocketResourceNameGivenSession(sessionId);
+
+    // Return a WebSocket URI constructed with host set to listener’s host, port set to listener’s port, path set to resource name, following the wss-URI construct if listener’s secure flag is set and the ws-URL construct otherwise.
+    // TODO: Support secure flag
+
+    auto host = listener->host;
+    if (host == "all"_s)
+        host = "localhost"_s;
+
+    return makeString("ws://"_s, host, ":"_s, String::number(listener->port), resourceName);
+}
+
+void WebSocketServer::removeResourceForSession(const String& sessionID)
+{
+    auto resourceName = constructWebSocketResourceNameGivenSession(sessionID);
+    m_listener->resources.erase(std::remove(m_listener->resources.begin(), m_listener->resources.end(), resourceName), m_listener->resources.end());
+}
+
+WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode errorCode, std::optional<Connection> connection, std::optional<String> errorMessage, std::optional<String> commandId)
+{
+    if (!connection)
+        return { };
+
+    auto reply = JSON::Object::create();
+
+    if (commandId)
+        reply->setString("id"_s, *commandId);
+
+    if (errorMessage)
+        reply->setString("message"_s, *errorMessage);
+
+    // TODO Refactor httpStatusCode as a static method in CommandResult?
```

Comment:
To remove?

Reply:
Removed the comment and refactored the actual code translation into a new `CommandResult::errorCodeToHTTPStatusCode`, also removing the temporary `CommandResult` used to get the code in the previous iteration.